napari/plugins/io.py

```diff
@@ -228,18 +230,40 @@ def save_layers(
     else:
         written = []
 
+    # If written is empty, something went wrong.
+    # Generate a warning to tell the user what it was.
     if not written:
-        # if written is empty, it means no plugin could write the
-        # path/layers combination
-        # we just want to provide some useful feedback
-        warnings.warn(
-            trans._(
-                'No data written! There may be no plugins capable of writing these {length} layers to {path}.',
-                deferred=True,
-                length=len(layers),
-                path=path,
+        if len(layers) == 0:
```

Comment:
any reason you can think of not to just raise an exception above `if not layers`?  might pop up in a test... but I can't think of any reason why we need to support calling this function with no layers

Reply:
I agree with @tlambert03 here - this should be moved up with the logic at the top to avoid unnecessary nested conditions.

Raising an exception and adding that to the docstring would be best, but I'd be fine with a warning for now if there's some reason to fear raising hear.

Unless there is some case where `save_layers` would support writing 0 layers? The only thing I can think of is a plugin that generates data that should be written immediately without going through a napari layer. But as I understand it, that's not a napari plugin writer contribution, and would be better served as a just a plain Python function.

Reply:
We do raise exceptions internally at various places, but we don't do it everywhere. We need to rework the return value conventions of these calls so it's clear.

I think that's out of scope for this PR. It's probably easiest to do this in #4692.